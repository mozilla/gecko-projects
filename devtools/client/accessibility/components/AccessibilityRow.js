/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
"use strict";

/* global gTelemetry, gToolbox, EVENTS */

// React & Redux
const {
  Component,
  createFactory,
} = require("devtools/client/shared/vendor/react");
const PropTypes = require("devtools/client/shared/vendor/react-prop-types");
const { findDOMNode } = require("devtools/client/shared/vendor/react-dom");
const { connect } = require("devtools/client/shared/vendor/react-redux");

const TreeRow = require("devtools/client/shared/components/tree/TreeRow");
const AuditFilter = createFactory(require("./AuditFilter"));
const AuditController = createFactory(require("./AuditController"));

// Utils
const {
  flashElementOn,
  flashElementOff,
} = require("devtools/client/inspector/markup/utils");
const { openDocLink } = require("devtools/client/shared/link");
const {
  VALUE_FLASHING_DURATION,
  VALUE_HIGHLIGHT_DURATION,
} = require("../constants");

// Actions
const { updateDetails } = require("../actions/details");
const { unhighlight } = require("../actions/accessibles");

const { L10N } = require("../utils/l10n");

loader.lazyRequireGetter(this, "Menu", "devtools/client/framework/menu");
loader.lazyRequireGetter(
  this,
  "MenuItem",
  "devtools/client/framework/menu-item"
);

const { scrollIntoView } = require("devtools/client/shared/scroll");

const JSON_URL_PREFIX = "data:application/json;charset=UTF-8,";

const TELEMETRY_ACCESSIBLE_CONTEXT_MENU_OPENED =
  "devtools.accessibility.accessible_context_menu_opened";
const TELEMETRY_ACCESSIBLE_CONTEXT_MENU_ITEM_ACTIVATED =
  "devtools.accessibility.accessible_context_menu_item_activated";

class HighlightableTreeRowClass extends TreeRow {
  shouldComponentUpdate(nextProps) {
    const shouldTreeRowUpdate = super.shouldComponentUpdate(nextProps);
    if (shouldTreeRowUpdate) {
      return shouldTreeRowUpdate;
    }

    if (
      nextProps.highlighted !== this.props.highlighted ||
      nextProps.filtered !== this.props.filtered
    ) {
      return true;
    }

    return false;
  }
}

const HighlightableTreeRow = createFactory(HighlightableTreeRowClass);

// Component that expands TreeView's own TreeRow and is responsible for
// rendering an accessible object.
class AccessibilityRow extends Component {
  static get propTypes() {
    return {
      ...TreeRow.propTypes,
      hasContextMenu: PropTypes.bool.isRequired,
      dispatch: PropTypes.func.isRequired,
      walker: PropTypes.object,
    };
  }

  componentDidMount() {
    const { selected, object } = this.props.member;
    if (selected) {
      this.unhighlight();
      this.update();
      this.highlight(object, { duration: VALUE_HIGHLIGHT_DURATION });
    }

    if (this.props.highlighted) {
      this.scrollIntoView();
    }
  }

  /**
   * Update accessible object details that are going to be rendered inside the
   * accessible panel sidebar.
   */
  componentDidUpdate(prevProps) {
    const { selected, object } = this.props.member;
    // If row is selected, update corresponding accessible details.
    if (!prevProps.member.selected && selected) {
      this.unhighlight();
      this.update();
      this.highlight(object, { duration: VALUE_HIGHLIGHT_DURATION });
    }

    if (this.props.highlighted) {
      this.scrollIntoView();
    }

    if (!selected && prevProps.member.value !== this.props.member.value) {
      this.flashValue();
    }
  }

  scrollIntoView() {
    const row = findDOMNode(this);
    // Row might not be rendered in the DOM tree if it is filtered out during
    // audit.
    if (!row) {
      return;
    }

    scrollIntoView(row);
  }

  update() {
    const {
      dispatch,
      member: { object },
      supports,
    } = this.props;
    if (!gToolbox || !object.actorID) {
      return;
    }

    dispatch(updateDetails(gToolbox.walker, object, supports));
    window.emit(EVENTS.NEW_ACCESSIBLE_FRONT_SELECTED, object);
  }

  flashValue() {
    const row = findDOMNode(this);
    // Row might not be rendered in the DOM tree if it is filtered out during
    // audit.
    if (!row) {
      return;
    }

    const value = row.querySelector(".objectBox");

    flashElementOn(value);
    if (this._flashMutationTimer) {
      clearTimeout(this._flashMutationTimer);
      this._flashMutationTimer = null;
    }
    this._flashMutationTimer = setTimeout(() => {
      flashElementOff(value);
    }, VALUE_FLASHING_DURATION);
  }

  highlight(accessible, options) {
    const { walker, dispatch } = this.props;
    dispatch(unhighlight());

    if (!accessible || !walker) {
      return;
    }

    walker
      .highlightAccessible(accessible, options)
      .catch(error => console.warn(error));
  }

  unhighlight() {
    const { walker, dispatch } = this.props;
    dispatch(unhighlight());

    if (!walker) {
      return;
    }

    walker.unhighlight().catch(error => console.warn(error));
  }

  async printToJSON() {
    const { member, supports } = this.props;
    if (!supports.snapshot) {
      // Debugger server does not support Accessible actor snapshots.
      return;
    }

    if (gTelemetry) {
      gTelemetry.keyedScalarAdd(
        TELEMETRY_ACCESSIBLE_CONTEXT_MENU_ITEM_ACTIVATED,
        "print-to-json",
        1
      );
    }

    const snapshot = await member.object.snapshot();
    openDocLink(
      `${JSON_URL_PREFIX}${encodeURIComponent(JSON.stringify(snapshot))}`
    );
  }

  onContextMenu(e) {
    e.stopPropagation();
    e.preventDefault();

    if (!gToolbox) {
      return;
    }

    const menu = new Menu({ id: "accessibility-row-contextmenu" });
    const { supports } = this.props;

    if (supports.snapshot) {
      menu.append(
        new MenuItem({
          id: "menu-printtojson",
          label: L10N.getStr("accessibility.tree.menu.printToJSON"),
          click: () => this.printToJSON(),
        })
      );
    }

    menu.popup(e.screenX, e.screenY, gToolbox.doc);

    if (gTelemetry) {
      gTelemetry.scalarAdd(TELEMETRY_ACCESSIBLE_CONTEXT_MENU_OPENED, 1);
    }
  }

  /**
   * Render accessible row component.
   * @returns acecssible-row React component.
   */
  render() {
    const { member } = this.props;
    const props = {
      ...this.props,
      onContextMenu: this.props.hasContextMenu && (e => this.onContextMenu(e)),
      onMouseOver: () => this.highlight(member.object),
      onMouseOut: () => this.unhighlight(),
      key: `${member.path}-${member.active ? "active" : "inactive"}`,
    };

    return AuditController(
      {
        accessible: member.object,
      },
      AuditFilter({}, HighlightableTreeRow(props))
    );
  }
}

const mapStateToProps = ({ ui }) => ({
  supports: ui.supports,
});

module.exports = connect(
  mapStateToProps,
  null,
  null,
  { withRef: true }
)(AccessibilityRow);
