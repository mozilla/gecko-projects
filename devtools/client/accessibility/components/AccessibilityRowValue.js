/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

"use strict";

const {
  Component,
  createFactory,
} = require("devtools/client/shared/vendor/react");
const PropTypes = require("devtools/client/shared/vendor/react-prop-types");
const { span } = require("devtools/client/shared/vendor/react-dom-factories");

const { connect } = require("devtools/client/shared/vendor/react-redux");

const Badges = createFactory(require("./Badges"));
const AuditController = createFactory(require("./AuditController"));

const { REPS } = require("devtools/client/shared/components/reps/reps");
const { Grip } = REPS;
const Rep = createFactory(REPS.Rep);

class AccessibilityRowValue extends Component {
  static get propTypes() {
    return {
      member: PropTypes.shape({
        object: PropTypes.object,
      }).isRequired,
      supports: PropTypes.object.isRequired,
    };
  }

  render() {
    const {
      member,
      supports: { audit },
    } = this.props;

    return span(
      {
        role: "presentation",
      },
      Rep({
        ...this.props,
        defaultRep: Grip,
        cropLimit: 50,
      }),
      audit &&
        AuditController(
          {
            accessible: member.object,
          },
          Badges()
        )
    );
  }
}

const mapStateToProps = ({ ui: { supports } }) => {
  return { supports };
};

module.exports = connect(mapStateToProps)(AccessibilityRowValue);
