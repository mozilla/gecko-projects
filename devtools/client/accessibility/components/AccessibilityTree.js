/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
"use strict";

/* global EVENTS */

// React & Redux
const {
  Component,
  createFactory,
} = require("devtools/client/shared/vendor/react");
const PropTypes = require("devtools/client/shared/vendor/react-prop-types");
const { connect } = require("devtools/client/shared/vendor/react-redux");

const TreeView = createFactory(
  require("devtools/client/shared/components/tree/TreeView")
);
// Reps
const { MODE } = require("devtools/client/shared/components/reps/reps");

const { fetchChildren } = require("../actions/accessibles");

const { L10N } = require("../utils/l10n");
const { isFiltered } = require("../utils/audit");
const AccessibilityRow = createFactory(require("./AccessibilityRow"));
const AccessibilityRowValue = createFactory(require("./AccessibilityRowValue"));
const { Provider } = require("../provider");

const { scrollIntoView } = require("devtools/client/shared/scroll");

/**
 * Renders Accessibility panel tree.
 */
class AccessibilityTree extends Component {
  static get propTypes() {
    return {
      walker: PropTypes.object,
      dispatch: PropTypes.func.isRequired,
      accessibles: PropTypes.object,
      expanded: PropTypes.object,
      selected: PropTypes.string,
      highlighted: PropTypes.object,
      supports: PropTypes.object,
      filtered: PropTypes.bool,
    };
  }

  constructor(props) {
    super(props);

    this.onNameChange = this.onNameChange.bind(this);
    this.onReorder = this.onReorder.bind(this);
    this.onTextChange = this.onTextChange.bind(this);
    this.renderValue = this.renderValue.bind(this);
  }

  /**
   * Add accessibility walker front event listeners that affect tree rendering
   * and updates.
   */
  componentWillMount() {
    const { walker } = this.props;
    walker.on("reorder", this.onReorder);
    walker.on("name-change", this.onNameChange);
    walker.on("text-change", this.onTextChange);
    return null;
  }

  componentDidUpdate(prevProps) {
    // When filtering is toggled, make sure that the selected row remains in
    // view.
    if (this.props.filtered !== prevProps.filtered) {
      const selected = document.querySelector(".treeTable .treeRow.selected");
      if (selected) {
        scrollIntoView(selected, { center: true });
      }
    }

    window.emit(EVENTS.ACCESSIBILITY_INSPECTOR_UPDATED);
  }

  /**
   * Remove accessible walker front event listeners.
   */
  componentWillUnmount() {
    const { walker } = this.props;
    walker.off("reorder", this.onReorder);
    walker.off("name-change", this.onNameChange);
    walker.off("text-change", this.onTextChange);
  }

  /**
   * Handle accessible reorder event. If the accessible is cached and rendered
   * within the accessibility tree, re-fetch its children and re-render the
   * corresponding subtree.
   * @param {Object} accessible accessible object that had its subtree
   *                            reordered.
   */
  onReorder(accessible) {
    if (this.props.accessibles.has(accessible.actorID)) {
      this.props.dispatch(fetchChildren(accessible));
    }
  }

  /**
   * Handle accessible name change event. If the name of an accessible changes
   * and that accessible is cached and rendered within the accessibility tree,
   * re-fetch its parent's children and re-render the corresponding subtree.
   * @param {Object} accessible accessible object that had its name changed.
   * @param {Object} parent     optional parent accessible object. Note: if it
   *                            parent is not present, we assume that the top
   *                            level document's name has changed and use
   *                            accessible walker as a parent.
   */
  onNameChange(accessible, parent) {
    const { accessibles, walker, dispatch } = this.props;
    parent = parent || walker;

    if (
      accessibles.has(accessible.actorID) ||
      accessibles.has(parent.actorID)
    ) {
      dispatch(fetchChildren(parent));
    }
  }

  /**
   * Handle accessible text change (change/insert/remove) event. If the text of
   * an accessible changes and that accessible is cached and rendered within the
   * accessibility tree, re-fetch its children and re-render the corresponding
   * subtree.
   * @param  {Object} accessible  accessible object that had its child text
   *                              changed.
   */
  onTextChange(accessible) {
    const { accessibles, dispatch } = this.props;
    if (accessibles.has(accessible.actorID)) {
      dispatch(fetchChildren(accessible));
    }
  }

  renderValue(props) {
    return AccessibilityRowValue(props);
  }

  /**
   * Render Accessibility panel content
   */
  render() {
    const columns = [
      {
        id: "default",
        title: L10N.getStr("accessibility.role"),
      },
      {
        id: "value",
        title: L10N.getStr("accessibility.name"),
      },
    ];

    const {
      accessibles,
      dispatch,
      expanded,
      selected,
      highlighted: highlightedItem,
      supports,
      walker,
      filtered,
    } = this.props;

    // Historically, the first context menu item is snapshot function and it is available
    // for all accessible object.
    const hasContextMenu = supports.snapshot;

    const renderRow = rowProps => {
      const { object } = rowProps.member;
      const highlighted = object === highlightedItem;
      return AccessibilityRow(
        Object.assign({}, rowProps, {
          walker,
          hasContextMenu,
          highlighted,
          decorator: {
            getRowClass: function() {
              return highlighted ? ["highlighted"] : [];
            },
          },
        })
      );
    };
    const className = filtered ? "filtered" : undefined;

    return TreeView({
      object: walker,
      mode: MODE.SHORT,
      provider: new Provider(accessibles, filtered, dispatch),
      columns: columns,
      className,
      renderValue: this.renderValue,
      renderRow,
      label: L10N.getStr("accessibility.treeName"),
      header: true,
      expandedNodes: expanded,
      selected,
      onClickRow(nodePath, event) {
        if (event.target.classList.contains("theme-twisty")) {
          this.toggle(nodePath);
        }

        this.selectRow(
          this.rows.find(row => row.props.member.path === nodePath),
          { preventAutoScroll: true }
        );
      },
      onContextMenuTree:
        hasContextMenu &&
        function(e) {
          // If context menu event is triggered on (or bubbled to) the TreeView, it was
          // done via keyboard. Open context menu for currently selected row.
          let row = this.getSelectedRow();
          if (!row) {
            return;
          }

          row = row.getWrappedInstance();
          row.onContextMenu(e);
        },
    });
  }
}

const mapStateToProps = ({
  accessibles,
  ui: { expanded, selected, supports, highlighted },
  audit: { filters },
}) => ({
  accessibles,
  expanded,
  selected,
  supports,
  highlighted,
  filtered: isFiltered(filters),
});
// Exports from this module
module.exports = connect(mapStateToProps)(AccessibilityTree);
