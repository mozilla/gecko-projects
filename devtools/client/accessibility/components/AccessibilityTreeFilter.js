/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
"use strict";

/* global gTelemetry */

// React
const { createFactory, Component } = require("devtools/client/shared/vendor/react");
const { div, span } = require("devtools/client/shared/vendor/react-dom-factories");
const PropTypes = require("devtools/client/shared/vendor/react-prop-types");
const { L10N } = require("../utils/l10n");
const ToggleButton = createFactory(require("./Button").ToggleButton);

const actions = require("../actions/audit");

const { connect } = require("devtools/client/shared/vendor/react-redux");
const { FILTERS } = require("../constants");

const TELEMETRY_AUDIT_ACTIVATED = "devtools.accessibility.audit_activated";
const FILTER_LABELS = {
  [FILTERS.CONTRAST]: "accessibility.badge.contrast",
};

class AccessibilityTreeFilter extends Component {
  static get propTypes() {
    return {
      auditing: PropTypes.string.isRequired,
      filters: PropTypes.object.isRequired,
      dispatch: PropTypes.func.isRequired,
      walker: PropTypes.object.isRequired,
      describedby: PropTypes.string,
    };
  }

  async toggleFilter(filterKey) {
    const { dispatch, filters, walker } = this.props;

    if (!filters[filterKey]) {
      if (gTelemetry) {
        gTelemetry.keyedScalarAdd(TELEMETRY_AUDIT_ACTIVATED, filterKey, 1);
      }

      dispatch(actions.auditing(filterKey));
      await dispatch(actions.audit(walker, filterKey));
    }

    // We wait to dispatch filter toggle until the tree is ready to be filtered
    // right after the audit. This is to make sure that we render an empty tree
    // (filtered) while the audit is running.
    dispatch(actions.filterToggle(filterKey));
  }

  onClick(filterKey, e) {
    const { mozInputSource, MOZ_SOURCE_KEYBOARD } = e.nativeEvent;
    if (e.isTrusted && mozInputSource === MOZ_SOURCE_KEYBOARD) {
      // Already handled by key down handler on user input.
      return;
    }

    this.toggleFilter(filterKey);
  }

  onKeyDown(filterKey, e) {
    // We explicitely handle "click" and "keydown" events this way here because
    // of the expectation of both Space and Enter triggering the click event
    // even though Enter is the only one in the spec.
    if (![" ", "Enter"].includes(e.key)) {
      return;
    }

    this.toggleFilter(filterKey);
  }

  render() {
    const { auditing, filters, describedby } = this.props;
    const toolbarLabelID = "accessibility-tree-filters-label";
    const filterButtons = Object.entries(filters).map(([filterKey, active]) =>
      ToggleButton({
        className: "badge",
        key: filterKey,
        active,
        label: L10N.getStr(FILTER_LABELS[filterKey]),
        onClick: this.onClick.bind(this, filterKey),
        onKeyDown: this.onKeyDown.bind(this, filterKey),
        busy: auditing === filterKey,
      }));

    return div({
      role: "toolbar",
      className: "accessibility-tree-filters",
      "aria-labelledby": toolbarLabelID,
      "aria-describedby": describedby,
    },
      span({ id: toolbarLabelID, role: "presentation" },
        L10N.getStr("accessibility.tree.filters")),
      ...filterButtons);
  }
}

const mapStateToProps = ({ audit: { filters, auditing } }) => {
  return { filters, auditing };
};

// Exports from this module
module.exports = connect(mapStateToProps)(AccessibilityTreeFilter);
