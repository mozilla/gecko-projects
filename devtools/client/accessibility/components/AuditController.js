/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

"use strict";

const React = require("devtools/client/shared/vendor/react");
const PropTypes = require("devtools/client/shared/vendor/react-prop-types");

class AuditController extends React.Component {
  static get propTypes() {
    return {
      accessible: PropTypes.object.isRequired,
      children: PropTypes.any,
    };
  }

  constructor(props) {
    super(props);

    const {
      accessible: { checks },
    } = props;
    this.state = {
      checks,
    };

    this.onAudited = this.onAudited.bind(this);
  }

  componentWillMount() {
    const { accessible } = this.props;
    accessible.on("audited", this.onAudited);
  }

  componentDidMount() {
    this.maybeRequestAudit();
  }

  componentDidUpdate() {
    this.maybeRequestAudit();
  }

  componentWillUnmount() {
    const { accessible } = this.props;
    accessible.off("audited", this.onAudited);
  }

  onAudited() {
    const { accessible } = this.props;
    if (!accessible.actorID) {
      // Accessible is being removed, stop listening for 'audited' events.
      accessible.off("audited", this.onAudited);
      return;
    }

    this.setState({ checks: accessible.checks });
  }

  maybeRequestAudit() {
    const { accessible } = this.props;
    if (!accessible.actorID) {
      // Accessible is being removed, stop listening for 'audited' events.
      accessible.off("audited", this.onAudited);
      return;
    }

    if (accessible.checks) {
      return;
    }

    accessible.audit().catch(error => {
      // Accessible actor was destroyed, connection closed.
      if (accessible.actorID) {
        console.warn(error);
      }
    });
  }

  render() {
    const { children } = this.props;
    const { checks } = this.state;

    return React.Children.only(React.cloneElement(children, { checks }));
  }
}

module.exports = AuditController;
