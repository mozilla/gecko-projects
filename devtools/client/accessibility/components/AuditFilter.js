/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

"use strict";

const React = require("devtools/client/shared/vendor/react");
const PropTypes = require("devtools/client/shared/vendor/react-prop-types");
const { connect } = require("devtools/client/shared/vendor/react-redux");

const { isFiltered } = require("../utils/audit");
const { FILTERS } = require("../constants");
const { accessibility: { AUDIT_TYPE, SCORES } } = require("devtools/shared/constants");

function validateContrast({ error, score }) {
  return !error && score === SCORES.FAIL;
}

const AUDIT_TYPE_TO_FILTER = {
  [AUDIT_TYPE.CONTRAST]: {
     filterKey: FILTERS.CONTRAST,
     validator: validateContrast,
  },
};

class AuditFilter extends React.Component {
  static get propTypes() {
    return {
      checks: PropTypes.object,
      children: PropTypes.any,
      filters: PropTypes.object.isRequired,
    };
  }

  isVisible(filters) {
    return !isFiltered(filters);
  }

  shouldHide() {
    const { filters, checks } = this.props;
    if (this.isVisible(filters)) {
      return false;
    }

    if (!checks || Object.values(checks).every(check => check == null)) {
      return true;
    }

    for (const type in checks) {
      if (AUDIT_TYPE_TO_FILTER[type] &&
          filters[AUDIT_TYPE_TO_FILTER[type].filterKey] &&
          AUDIT_TYPE_TO_FILTER[type].validator(checks[type])) {
        return false;
      }
    }

    return true;
  }

  render() {
    return this.shouldHide() ? null : this.props.children;
  }
}

const mapStateToProps = ({ audit: { filters } }) => {
  return { filters };
};

module.exports = connect(mapStateToProps)(AuditFilter);
