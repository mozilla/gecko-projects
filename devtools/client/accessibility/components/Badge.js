/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
"use strict";

// React
const { Component, createFactory } = require("devtools/client/shared/vendor/react");
const PropTypes = require("devtools/client/shared/vendor/react-prop-types");
const { connect } = require("devtools/client/shared/vendor/react-redux");

const ToggleButton = createFactory(require("./Button").ToggleButton);

const { audit, auditing, filterToggle } = require("../actions/audit");
const { preventDefaultAndStopPropagation } = require("devtools/client/shared/events");

class Badge extends Component {
  static get propTypes() {
    return {
      active: PropTypes.bool.isRequired,
      filterKey: PropTypes.string.isRequired,
      dispatch: PropTypes.func.isRequired,
      label: PropTypes.string.isRequired,
      tooltip: PropTypes.string,
      walker: PropTypes.object.isRequired,
    };
  }

  constructor(props) {
    super(props);

    this.toggleFilter = this.toggleFilter.bind(this);
    this.onClick = this.onClick.bind(this);
    this.onKeyDown = this.onKeyDown.bind(this);
  }

  shouldComponentUpdate(nextProps) {
    return nextProps.active !== this.props.active;
  }

  async toggleFilter() {
    const { dispatch, filterKey, walker, active } = this.props;
    if (!active) {
      dispatch(auditing(filterKey));
      await dispatch(audit(walker, filterKey));
    }

    // We wait to dispatch filter toggle until the tree is ready to be filtered
    // right after the audit. This is to make sure that we render an empty tree
    // (filtered) while the audit is running.
    dispatch(filterToggle(filterKey));
  }

  onClick(e) {
    preventDefaultAndStopPropagation(e);
    const { mozInputSource, MOZ_SOURCE_KEYBOARD } = e.nativeEvent;
    if (e.isTrusted && mozInputSource === MOZ_SOURCE_KEYBOARD) {
      // Already handled by key down handler on user input.
      return;
    }

    this.toggleFilter();
  }

  onKeyDown(e) {
    // We explicitely handle "click" and "keydown" events this way here because
    // there seem to be a difference in the sequence of keyboard/click events
    // fired when Space/Enter is pressed. When Space is pressed the sequence of
    // events is keydown->keyup->click but when Enter is pressed the sequence is
    // keydown->click->keyup. This results in an unwanted badge click (when
    // pressing Space) within the accessibility tree row when activating it
    // because it gets focused before the click event is dispatched.
    if (![" ", "Enter"].includes(e.key)) {
      return;
    }

    preventDefaultAndStopPropagation(e);
    this.toggleFilter();
  }

  render() {
    const { active, label, tooltip } = this.props;

    return ToggleButton({
      className: "audit-badge badge",
      label,
      active,
      tooltip,
      onClick: this.onClick,
      onKeyDown: this.onKeyDown,
    });
  }
}

const mapStateToProps = ({ audit: { filters } }, { filterKey }) => ({
  active: filters[filterKey],
});

module.exports = connect(mapStateToProps)(Badge);
