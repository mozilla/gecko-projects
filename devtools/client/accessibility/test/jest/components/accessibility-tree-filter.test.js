/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
"use strict";

const { mount } = require("enzyme");

const { createFactory } = require("devtools/client/shared/vendor/react");
const Provider = createFactory(require("devtools/client/shared/vendor/react-redux").Provider);

const { ToggleButton } = require("devtools/client/accessibility/components/Button");
const ConnectedAccessibilityTreeFilterClass =
  require("devtools/client/accessibility/components/AccessibilityTreeFilter");
const AccessibilityTreeFilterClass =
  ConnectedAccessibilityTreeFilterClass.WrappedComponent;
const AccessibilityTreeFilter = createFactory(ConnectedAccessibilityTreeFilterClass);
const {
  setupStore,
} = require("devtools/client/accessibility/test/jest/helpers");
const { FILTERS } = require("devtools/client/accessibility/constants");

describe("AccessibilityTreeFilter component:", () => {
  it("audit filter not filtered", () => {
    const store = setupStore();

    const wrapper = mount(Provider({store}, AccessibilityTreeFilter()));
    expect(wrapper.html()).toMatchSnapshot();

    const filters = wrapper.find(AccessibilityTreeFilterClass);
    expect(filters.children().length).toBe(1);

    const toolbar = filters.childAt(0);
    expect(toolbar.is("div")).toBe(true);
    expect(toolbar.prop("role")).toBe("toolbar");

    const filterButtons = filters.find(ToggleButton);
    expect(filterButtons.length).toBe(1);

    const button = filterButtons.at(0).childAt(0);
    expect(button.is("button")).toBe(true);
    expect(button.hasClass("audit-badge")).toBe(true);
    expect(button.hasClass("badge")).toBe(true);
    expect(button.hasClass("toggle-button")).toBe(true);
    expect(button.prop("aria-pressed")).toBe(false);
    expect(button.text()).toBe("accessibility.badge.contrast");
  });

  it("audit filter filtered", () => {
    const store = setupStore({
      preloadedState: { audit: { filters: { [FILTERS.CONTRAST]: true }}},
    });

    const wrapper = mount(Provider({store}, AccessibilityTreeFilter()));
    expect(wrapper.html()).toMatchSnapshot();

    const button = wrapper.find("button");
    expect(button.prop("aria-pressed")).toBe(true);
    expect(button.hasClass("checked")).toBe(true);
  });

  it("audit filter not filtered auditing", () => {
    const store = setupStore({
      preloadedState: { audit: {
        filters: {
          [FILTERS.CONTRAST]: false,
        },
        auditing: FILTERS.CONTRAST,
      }},
    });

    const wrapper = mount(Provider({store}, AccessibilityTreeFilter()));
    expect(wrapper.html()).toMatchSnapshot();

    const button = wrapper.find("button");
    expect(button.prop("aria-pressed")).toBe(false);
    expect(button.hasClass("checked")).toBe(false);
    expect(button.prop("aria-busy")).toBe(true);
    expect(button.hasClass("devtools-throbber")).toBe(true);
  });

  it("audit filter filtered auditing", () => {
    const store = setupStore({
      preloadedState: { audit: {
        filters: {
          [FILTERS.CONTRAST]: true,
        },
        auditing: FILTERS.CONTRAST,
      }},
    });

    const wrapper = mount(Provider({store}, AccessibilityTreeFilter()));
    expect(wrapper.html()).toMatchSnapshot();

    const button = wrapper.find("button");
    expect(button.prop("aria-pressed")).toBe(true);
    expect(button.hasClass("checked")).toBe(true);
    expect(button.prop("aria-busy")).toBe(true);
    expect(button.hasClass("devtools-throbber")).toBe(true);
  });

  it("toggle filter", () => {
    const store = setupStore();
    const wrapper = mount(Provider({store}, AccessibilityTreeFilter()));
    expect(wrapper.html()).toMatchSnapshot();

    const filterInstance = wrapper.find(AccessibilityTreeFilterClass).instance();
    filterInstance.toggleFilter = jest.fn();
    wrapper.find("button.audit-badge.badge").simulate("keydown", { key: " " });
    expect(filterInstance.toggleFilter.mock.calls.length).toBe(1);

    wrapper.find("button.audit-badge.badge").simulate("keydown", { key: "Enter" });
    expect(filterInstance.toggleFilter.mock.calls.length).toBe(2);

    wrapper.find("button.audit-badge.badge").simulate("click", { clientX: 1 });
    expect(filterInstance.toggleFilter.mock.calls.length).toBe(3);
  });
});
