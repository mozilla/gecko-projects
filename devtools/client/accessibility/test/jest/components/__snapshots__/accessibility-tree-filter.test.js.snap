// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccessibilityTreeFilter component: audit filter filtered 1`] = `"<div role=\\"toolbar\\" class=\\"accessibility-tree-filters\\" aria-labelledby=\\"accessibility-tree-filters-label\\"><span id=\\"accessibility-tree-filters-label\\" role=\\"presentation\\">accessibility.tree.filters</span><button aria-pressed=\\"true\\" aria-busy=\\"false\\" class=\\"badge toggle-button checked\\">accessibility.badge.contrast</button></div>"`;

exports[`AccessibilityTreeFilter component: audit filter filtered auditing 1`] = `"<div role=\\"toolbar\\" class=\\"accessibility-tree-filters\\" aria-labelledby=\\"accessibility-tree-filters-label\\"><span id=\\"accessibility-tree-filters-label\\" role=\\"presentation\\">accessibility.tree.filters</span><button aria-pressed=\\"true\\" aria-busy=\\"true\\" class=\\"badge toggle-button checked devtools-throbber\\">accessibility.badge.contrast</button></div>"`;

exports[`AccessibilityTreeFilter component: audit filter not filtered 1`] = `"<div role=\\"toolbar\\" class=\\"accessibility-tree-filters\\" aria-labelledby=\\"accessibility-tree-filters-label\\"><span id=\\"accessibility-tree-filters-label\\" role=\\"presentation\\">accessibility.tree.filters</span><button aria-pressed=\\"false\\" aria-busy=\\"false\\" class=\\"badge toggle-button\\">accessibility.badge.contrast</button></div>"`;

exports[`AccessibilityTreeFilter component: audit filter not filtered auditing 1`] = `"<div role=\\"toolbar\\" class=\\"accessibility-tree-filters\\" aria-labelledby=\\"accessibility-tree-filters-label\\"><span id=\\"accessibility-tree-filters-label\\" role=\\"presentation\\">accessibility.tree.filters</span><button aria-pressed=\\"false\\" aria-busy=\\"true\\" class=\\"badge toggle-button devtools-throbber\\">accessibility.badge.contrast</button></div>"`;

exports[`AccessibilityTreeFilter component: toggle filter 1`] = `"<div role=\\"toolbar\\" class=\\"accessibility-tree-filters\\" aria-labelledby=\\"accessibility-tree-filters-label\\"><span id=\\"accessibility-tree-filters-label\\" role=\\"presentation\\">accessibility.tree.filters</span><button aria-pressed=\\"false\\" aria-busy=\\"false\\" class=\\"badge toggle-button\\">accessibility.badge.contrast</button></div>"`;
